### PR TITLE
First Draft

### PR DESCRIPTION
**Pull request by joshua-a-becker**
https://github.com/empiricaly/meteor-empirica-core/issues/32

### CODE REVIEW

**Comment by joshua-a-becker** [comment]
added node_modules to .gitignore

File: client/game/SocialExposure.jsx [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+    const { game, player } = this.props;
+
+    const otherPlayers = _.filter(
+      game.players,
+      p =>
+        p._id != player._id && player.get("neighbors").includes(p.get("nodeId"))
+    );
+
+    window.player = player;
+    window.op = otherPlayers;
```
Comment (npaton) [inline client/game/SocialExposure.jsx]: I think these `windows.variable` references are for debugging, right? I don't think we should keep this in released code. It might be a practical way to debug, but not something in the scope of the intro tutorial.

File: client/game/SocialExposure.jsx [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+    );
+  }
+
+  render() {
+    const { game, player } = this.props;
+
+    const otherPlayers = _.filter(
+      game.players,
+      p =>
+        p._id != player._id && player.get("neighbors").includes(p.get("nodeId"))
```
Comment (npaton) [inline client/game/SocialExposure.jsx]: This arrow function might be kinda hard to parse for a beginner. I think "neighbors" should not contain the current user from the get go. Maybe also we can pull out `player.get("neighbors")` into a var above `const otherPlayers`

Would give something like:

```suggestion
        neighbors.includes(otherPlayer.get("nodeId"))
```

File: client/game/SocialExposure.jsx [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+    }
+
+    return (
+      <div className="social-exposure">
+        <p>
+          <strong>There are {otherPlayers.length} other players:</strong>
+        </p>
+        {!_.isEmpty(otherPlayers)
+          ? otherPlayers.map(p => this.renderSocialInteraction(p))
+          : ""}
```
Comment (npaton) [inline client/game/SocialExposure.jsx]: The `!_.isEmpty(otherPlayers) ? ... : ...` check is not necessary, as the 0 other players condition is capture on line 27:
```js
    if (otherPlayers.length === 0) {
      return null;
    }
```

File: client/game/TaskStimulus.jsx [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+
+    window.round = round;
+
+    const imagePath = round.get("imagePath");
+    const questionText = round.get("questionText");
+
+    return (
+      <div className="task-stimulus">
+        <div className="task-image">
+          {imagePath == undefined ? (
```
Comment (npaton) [inline client/game/TaskStimulus.jsx]:
```suggestion
          {imagePath ? (
```

If `imagePath` is undefined, it is also false. Not comparing to undefined will also have the added advantage of catching other unfortunate cases of bad values such as "" (empty string) or null.

File: client/game/TaskStimulus.jsx [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+            ""
+          ) : (
+            <img src={imagePath} height={"300px"} />
```
Comment (npaton) [inline client/game/TaskStimulus.jsx]: It would be preferable here to use CSS over the `"height"` attribute. We already have a class surrounding this element that we can use:

```css
.task-image img {
    height: 300px;
}
```

File: client/main.js [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+import Round from "./game/Round";
+import Consent from "./intro/Consent";
+import InstructionStepOne from "./intro/InstructionStepOne";
+import InstructionStepTwo from "./intro/InstructionStepTwo";
+import Quiz from "./intro/Quiz";
+import ExitSurvey from "./exit/ExitSurvey";
+import Thanks from "./exit/Thanks";
+
+// Set the Consent Component you want to present players (optional).
+Empirica.consent();
```
Comment (npaton) [inline client/main.js]: This should generally be commented out if you're not changing it. In this case, I think it maybe should be:

```suggestion
Empirica.consent(Consent);
```

...since `Consent` from "./intro/Consent" is loaded into this file.

File: local.json [hunk 7]
```diff
@@ -0,0 +1,8 @@
+{
+  "admins": [
+    {
+      "username": "admin",
+      "password": "h1hi25ojw4"
+    }
+  ]
+}
```
Comment (npaton) [inline local.json]: I don't think this file should be part of this repo, it contradicts the tutorial that says the admin password will be in the console. It will be as people will probably not use `meteor --settings local.json` on their own. So having this file becomes more confusing than anything else. Also, if people use this repo as their starting point to create their own project, they might not think to update this file and inadvertently use this password for their project.

File: server/main.js [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+// gameInit is where the structure of a game is defined.
+// Just before every game starts, once all the players needed are ready, this
+// function is called with the treatment and the list of players.
+// You must then add rounds and stages to the game, depending on the treatment
+// and the players. You can also get/set initial values on your game, players,
+// rounds and stages (with get/set methods), that will be able to use later in
+// the game.
+Empirica.gameInit(game => {
+  // Establish node list
+  console.log(game);
```
Comment (npaton) [inline server/main.js]: We should remove all `console.log` appearances. They are easy to add back, and will clutter the console with info the user will not know where it's coming from. Also, make the comment here a bit misplaced.

File: server/main.js [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+
+  const nodes = [];
+  for (var i = game.players.length; i--; i > 0) nodes.push(i);
```
Comment (npaton) [inline server/main.js]: This cannot have been prettified, single line loop not supported.

That said, I am not sure about this whole neighbors scheme. In its current form, you could end up with no neighbors, while adding the current player as a neighbor. While both of these situations are caught in the UI, they both lead to sub-optimal situations. Without making this more complex, I think we can do this neighbors stuff differently.

I propose using the actual player IDs (player._id). We don't need the `nodes` var here anymore, nor do we need the "nodeId" piece of data on each player, and for the `networkNeighbors ` we can do something like the following. Before the big `game.players.forEach` loop, we collect all player IDs:

```js
const playerIds = game.players.forEach(player => player._id);
```

Then for each player:

```js
const otherPlayers = _.without(playerIds, player._id);
player.set("neighbors", _.sample(otherPlayers, 3));
```

[`_.sample`](https://underscorejs.org/#sample) does kinda guaranty we have neighbors. Unless there are 1 or 2 players. We could make the sample size (here 3), dynamic, like `_.max(0, otherplayers - 2)`, but it does make things more complex, maybe uselessly for an intro tutorial.

**Review by npaton — CHANGES_REQUESTED** [review]
I think it's pretty cool, and not too far from done imo. My comments are mostly cosmetic. Since it's meant to be one of the first pieces of code people see when starting Empirica, being consistent and clean is important I think. But yeah, not much to say, it's nearly there.

The one thing that I would like to change if possible is the neighbors. I made a large comment near the `gameInit` explaining the idea (using player IDs), but changes elsewhere would be needed for it to work and I did not detail these changes, though they are not many, and it should be a relatively easy fix.

File: client/game/SocialExposure.jsx (hunk 1 above)
Comment (joshua-a-becker) [reply client/game/SocialExposure.jsx]: oh whoops yes, agreed

File: client/game/SocialExposure.jsx (hunk 2 above)
Comment (joshua-a-becker) [reply client/game/SocialExposure.jsx]: agreed re: neighbors doesn't contain player, this is some weird Frankenstein leftovers from when i copied abdullah's tutorial code and turned it into network code

agreed re: arrow

File: server/main.js (hunk 8 above)
Comment (joshua-a-becker) [reply server/main.js]: yes sorry!

File: server/main.js (hunk 9 above)
Comment (joshua-a-becker) [reply server/main.js]: -- i am not worried about the possibility of zero neighbors, BUT
-- i like the idea of sampling from other players
-- the reason i had defined 'nodes' was to set people on a path to a more sustainable way of doing it which (imho) is to pass in a graph defined elsewhere, rather than generate the graph in the javascript (among other reasons, so that it's not stochastic)
-- but this is definitely much cleaner for a demo,  and someone thinking about that would already know how to do it

**Comment by joshua-a-becker** [comment]
great i'll make these revisions by end of tomorrow at the latest, probably today assuming i can get my other project done first

File: client/game/SocialExposure.jsx (hunk 2 above)
Comment (npaton) [reply client/game/SocialExposure.jsx]: Yeah, this one actually changes completely with moving neighbors to player IDs, my bad for the not updating this comment after I made that proposal below.

File: server/main.js (hunk 9 above)
Comment (npaton) [reply server/main.js]: Totally agreed, networks should not be done like this eventually. Not sure we talked about this already, but we have tried to add some sort of simple network support into Empirica and realized we needed something better than what we had planned, we were really falling short. But we do eventually want to give some cool tools for game designers to generate and use networks. One day... 💃

Comment (joshua-a-becker) [reply server/main.js]: ah well i'd be happy to contribute to that effort!

a conversation for another page, but:  a very simple approach is to put a large TextArea input (as a config) that lets people put in an edgelist  (each row is a user, the first entry is their node, and the remaining entries are their neighbors, all separated by spaces or commas)

Comment (npaton) [reply server/main.js]: For sure. That kind of thing can help create the graph. I can't recall all we wanted to do, but I remember we also wanted things like once the graph exists, you might want to edit it (say between rounds, depending on conditions), or you want to be able to traverse it (i.e. friends or friends), etc.

We talk about it again when we get more core stuff smoothed out. ⛵️

**Comment by joshua-a-becker** [comment]
er, by end of today.

File: client/game/SocialExposure.jsx (hunk 1 above)
Comment (joshua-a-becker) [reply client/game/SocialExposure.jsx]: pushed

File: client/game/SocialExposure.jsx (hunk 3 above)
Comment (joshua-a-becker) [reply client/game/SocialExposure.jsx]: i'm reluctant to make this change here first, because it's an edit from the template---and thus it requires more work of the person following the demo

should be removed from the template first

Comment (joshua-a-becker) [reply client/game/SocialExposure.jsx]: i seem to be unable to start an issue for create-empirica-app

i would be happy to submit a pull request but, i don't know stylistically which of the redundancies you want to remove

File: client/game/TaskStimulus.jsx (hunk 4 above)
Comment (joshua-a-becker) [reply client/game/TaskStimulus.jsx]: TIL

File: client/game/TaskStimulus.jsx (hunk 5 above)
Comment (joshua-a-becker) [reply client/game/TaskStimulus.jsx]: this update is pushed and i'll update the demo code to match

File: server/main.js (hunk 9 above)
Comment (joshua-a-becker) [reply server/main.js]: this update is pushed

File: local.json (hunk 7 above)
Comment (joshua-a-becker) [reply local.json]: this is part of the create-empirica-app template so i'm not sure how to proceed on this part---it'll be there if they follow the instructions.

**Comment by joshua-a-becker** [comment]
i believe i've made all the updates as discussed above, except the two that would produce inconsistencies with the template

File: client/game/SocialExposure.jsx (hunk 3 above)
Comment (npaton) [reply client/game/SocialExposure.jsx]: Oh my, I did not realize this was part of the default app, my bad! I created a PR showing the change: https://github.com/empiricaly/create-empirica-app/pull/8/files. You can't merge it back into this project, but you can see what I mean this way I think. 💃

File: local.json (hunk 7 above)
Comment (npaton) [reply local.json]: You should be able to simply:
```sh
git rm local.json
git commit -m "Remove local.json"
```

Comment (joshua-a-becker) [reply local.json]: right, the issue is that this is part of the template, so the file will be there if they follow the instructions---we want the instructions to get them to a repo that matches the demo repo.

File: client/game/SocialExposure.jsx (hunk 3 above)
Comment (joshua-a-becker) [reply client/game/SocialExposure.jsx]: going to bed now, will do tomorrow

File: local.json (hunk 7 above)
Comment (npaton) [reply local.json]: Right, but we don't want to spread a password through this repo. Most people should go through the tutorial and will have a custom local.json generated. But for those who decide to clone this repo, we don't want them to have a "weak" password.

Event when local.json is properly generated, we don't really encourage to do `meteor --settings local.json`. We should, and I've been holding off. My aspiration here is to work on better configuration, this whole admin password thing printing to stderr is sketchy. What I want to do is to have a YAML or JS file somewhere (probably `/empirica.js` or `/empirica.yaml`) with all the Empirica specific configuration (as opposed to the Meteor specific stuff), that would be generated on startup if not present, so we'd be sure it's there, and the password wouldn't change on each reload as it does now if you don't specify the local.json. This upcoming change will fix this and more (all existing configuration would move there, and be well commented and structured).

Comment (joshua-a-becker) [reply local.json]: ah i understand sorry for my density and my delay---this has been pushed!

File: client/game/SocialExposure.jsx (hunk 3 above)
Comment (joshua-a-becker) [reply client/game/SocialExposure.jsx]: this update is pushed!

**Comment by joshua-a-becker** [comment]
i think that resolves all the open issues?

**Review by npaton — APPROVED** [review]

**Comment by npaton** [comment]
Thank you @joshua-a-becker !!  🎉
